wgsl/index.bs

```diff
@@ -2544,51 +2547,103 @@ programmable parts of pipelines:
+the [SHORTNAME] module and the entry point's function name.
 
-<pre class='def'>
-entry_point_decl
-  : ENTRY_POINT pipeline_stage EQUAL IDENT
-  | ENTRY_POINT pipeline_stage AS STRING_LITERAL EQUAL IDENT
-  | ENTRY_POINT pipeline_stage AS IDENT EQUAL IDENT
-</pre>
+An entry point function must have no parameters,
+and its return type must be [=void=].
+
+ISSUE: We've lost the ability to use shader names that are invalid identifiers.
+Do Metal and D3D12 limit entry point names to valid identifiers in C?
 
 <div class='example' heading='Entry Point'>
   <xmp>
-    entry_point vertex = main
-       OpEntryPoint Vertex %vtx_main "vtx_main" %gl_FragColor
+    [[stage vertex]]
+    fn vtx_main() -> void { gl_Position = vec4<f32>(); }
+       # OpEntryPoint Vertex %vtx_main "vtx_main" %gl_Position
 
-    entry_point fragment as “frag_main” = main
-       OpEntryPoint Fragment %main "frag_main" %gl_FragColor
+    [[stage fragment]]
+    fn frag_main -> void { gl_FragColor = vec4<f32>(); }
+       # OpEntryPoint Fragment %frag_main "frag_main" %gl_FragColor
 
-    entry_point compute = comp_main
-       OpEntryPoint GLCompute %comp_main "comp_main" %gl_FragColor
+    [[stage compute]]
+    fn main() -> void { }
+       # OpEntryPoint GLCompute %main "main"
   </xmp>
 </div>
+
+The set of <dfn noexport>functions in a shader stage</dfn> is the union of:
+
+* The entry point function for the stage.
+* The targets of function calls from within the body of a function
+    in the shader stage, whether or not that call is executed.
+
+The union is applied repeatedly until it stabilizes.
+It will stabilize in a finite number of steps.
+
+### Function attributes for entry points ### {#entry-point-attributes}
+
+<pre class='def'>
+function_decoration
+  : STAGE pipeline_stage
+  | WORKGROUP_SIZE INT_LITERAL (INT_LITERAL INT_LITERAL?)?
+
+pipeline_stage
+  : COMPUTE
+  | VERTEX
+  | FRAGMENT
+</pre>
+
+: <dfn noexport>stage</dfn>
+:: The `stage` attribute declares that a function is an entry point for particular pipeline stage.
+: <dfn noexport>workgroup_size</dfn>
+:: The `workgroup_size` attribute specifies the x, y, and z dimensions of the [=workgroup grid=]
+    for a [=compute=] shader.
+    The size in the x dimension is provided by the first literal.
+    The size in the y dimension is provided by the second literal, when present, and otherwise
+    is assumed to be 1.
+    The size in the z dimension is provided by the third literal, when present, and otherwise
+    is assumed to be 1.
+    Each dimension size must be at least 1 and at most an upper bound specified by the WebGPU API.
+    This attribute must only be used with a [=compute shader stage=] entry point.
+
+ISSUE: Can we query upper bounds on workgroup size dimensions?  Is it independent of the shader, or
+  a property to be queried after creating the shader module?
+
+<div class='example' heading='workgroup_size Attribute'>
+  <xmp>
+    [[ stage compute, workgroup_size 8 1 1 ]]
```

Comment:
Reading this, leans me in favour of `[[stage(compute), workgroup_size(8, 1, 1)]]`. More typing by clearer.

Reply:
Yes, that's the resolution from the [2020-09-15](https://docs.google.com/document/d/1FbHuoLT7Pmklgjx4_09KOCOFHQGwrfVXGvaoGytIOsU/edit#heading=h.22uy73bqa3qv) meeting